napari/layers/_shapes_layer/model.py

```diff
@@ -0,0 +1,1521 @@
+            elif (self.mode == 'direct' or self.mode == 'vertex_insert' or
+                  self.mode == 'vertex_remove'):
+                # Check if inside vertex of shape
+                inds = np.isin(self.data._index, self.selected_shapes)
+                vertices = self.data._vertices[inds]
+                distances = abs(vertices - coord[:2])
+
+                # Get the vertex sizes
+                scene = self.viewer._canvas.scene
+                transform = scene.node_transform(self._node)
+                rescale = transform.map([1, 1])[:2] - transform.map([0, 0])[:2]
+                sizes = self._vertex_size*rescale.mean()/2
+
+                # Check if any matching vertices
+                matches = np.all(distances <= sizes, axis=1).nonzero()[0]
+                if len(matches) > 0:
+                    index = inds.nonzero()[0][matches[-1]]
+                    shape = self.data._index[index]
+                    _, idx = np.unique(self.data._index, return_index=True)
+                    return shape, index - idx[shape]
+
+        # Check if mouse inside shape
+        triangles = self.data._mesh_vertices[self.data._mesh_triangles]
+        inside = inside_triangles(triangles - coord[:2])
+        shapes = self.data._mesh_triangles_index[inside]
+
+        if len(shapes) > 0:
+            indices = shapes[:, 0]
+            z_list = self.data._z_order.tolist()
+            order_indices = np.array([z_list.index(m) for m in indices])
+            ordered_shapes = indices[np.argsort(order_indices)]
+            return ordered_shapes[0], None
+        else:
+            return None, None
+
+    def _shapes_in_box(self, corners):
+        """Determines which shapes, if any, are inside an axis aligned box
+
+        Parameters
+        ----------
+        corners : np.ndarray
+            2x2 array of two corners that will be used to create an axis
+            aligned box.
+
+        Returns
+        ----------
+        shapes : list
+            List of shapes that are inside the box.
+        """
+
+        triangles = self.data._mesh_vertices[self.data._mesh_triangles]
+        intersects = triangles_intersect_box(triangles, corners)
+        shapes = self.data._mesh_triangles_index[intersects, 0]
+        shapes = np.unique(shapes).tolist()
+
+        return shapes
+
+    def _get_coord(self, position):
+        """Converts the position of the cursor in canvas coordinates to image
+        coordinates
+
+        Parameters
+        ----------
+        position : sequence of int
+            Position of mouse cursor in canvas coordinates.
+
+        Returns
+        ----------
+        coord : sequence of float
+            Position of mouse cursor in image coordinates.
+        """
+        transform = self.viewer._canvas.scene.node_transform(self._node)
+        pos = transform.map(position)
+        pos = [pos[1], pos[0]]
+        coord = np.array([pos[1], pos[0]])
```

Comment:
This code is super weird...? Why is pos inverted then inverted again?

Reply:
yup, no reason - fixed!